crates/sui-framework/sources/collectible/kiosk.move

```diff
@@ -0,0 +1,474 @@
+        self.item_count = self.item_count + 1;
+        dof::add(&mut self.id, Key { id: object::id(&item) }, item)
+    }
+
+    /// Take any object from the Kiosk.
+    /// Performs an authorization check to make sure only owner can do that.
+    public fun take<T: key + store>(
+        self: &mut Kiosk, cap: &KioskOwnerCap, id: ID
+    ): T {
+        assert!(object::id(self) == cap.for, ENotOwner);
+        self.item_count = self.item_count - 1;
+        df::remove_if_exists<Offer, u64>(&mut self.id, Offer { id });
+        dof::remove(&mut self.id, Key { id })
+    }
+
+    // === Trading functionality ===
+
+    /// Make an offer by setting a price for the item and making it publicly
+    /// purchasable by anyone on the network.
+    ///
+    /// Performs an authorization check to make sure only owner can sell.
+    public fun make_offer<T: key + store>(
+        self: &mut Kiosk, cap: &KioskOwnerCap, id: ID, price: u64
+    ) {
+        assert!(object::id(self) == cap.for, ENotOwner);
+        df::add(&mut self.id, Offer { id }, price);
+        event::emit(NewOfferEvent<T> {
+            kiosk: object::id(self), id, price
+        })
+    }
+
+    /// Place an item into the Kiosk and make an offer - simplifies the flow.
+    public fun place_and_offer<T: key + store>(
+        self: &mut Kiosk, cap: &KioskOwnerCap, item: T, price: u64
+    ) {
+        let id = object::id(&item);
+        place(self, cap, item);
+        make_offer<T>(self, cap, id, price)
+    }
+
+    /// Make a trade: pay the owner of the item and request a Transfer to the `target`
+    /// kiosk (to prevent item being taken by the approving party).
+    ///
+    /// Received `TransferRequest` needs to be handled by the publisher of the T,
+    /// if they have a method implemented that allows a trade, it is possible to
+    /// request their approval (by calling some function) so that the trade can be
+    /// finalized.
+    ///
+    /// After a confirmation is received from the creator, an item can be placed to
+    /// a destination safe.
+    public fun purchase<T: key + store>(
+        self: &mut Kiosk, id: ID, payment: Coin<SUI>
+    ): (T, TransferRequest<T>) {
+        let price = df::remove<Offer, u64>(&mut self.id, Offer { id });
+        let inner = dof::remove<Key, T>(&mut self.id, Key { id });
+
+        self.item_count = self.item_count - 1;
+        assert!(price == coin::value(&payment), EIncorrectAmount);
+        balance::join(&mut self.profits, coin::into_balance(payment));
+
+        (inner, TransferRequest<T> {
+            paid: price,
+            from: object::id(self),
+        })
+    }
+
+    /// Allow a `TransferRequest` for the type `T`. The call is protected
+    /// by the type constraint, as only the publisher of the `T` can get
+    /// `TransferPolicyCap<T>`.
+    ///
+    /// Note: unless there's a policy for `T` to allow transfers,
+    /// Kiosk trades will not be possible.
```

Comment:
nit: add some docs here to indicate that this function is expected to be used by the implementations of transfer policies, and point to an example one?